Tests/DataStreamTests.swift

```diff
@@ -0,0 +1,749 @@
+        var firstResponse: HTTPURLResponse?
+        var firstStreamOnMain = false
+        var firstCompleteOnMain = false
+        let firstExpectation = expectation(description: "first stream should complete")
+        var decodedResponse: HTTPBinResponse?
+        var decodingError: AFError?
+        var secondResponse: HTTPURLResponse?
+        var secondStreamOnMain = false
+        var secondCompleteOnMain = false
+        let secondExpectation = expectation(description: "second stream should complete")
+
+        // When
+        AF.streamRequest(URLRequest.makeHTTPBinRequest(path: "stream/1"))
+            .responseStream { stream in
+                switch stream.event {
+                case let .stream(result):
+                    firstStreamOnMain = Thread.isMainThread
+                    switch result {
+                    case let .success(data):
+                        firstAccumulatedData.append(data)
+                    }
+                case let .complete(completion):
+                    firstCompleteOnMain = Thread.isMainThread
+                    firstResponse = completion.response
+                    firstExpectation.fulfill()
+                }
+            }
+            .responseStreamDecodable(of: HTTPBinResponse.self) { stream in
+                switch stream.event {
+                case let .stream(result):
+                    secondStreamOnMain = Thread.isMainThread
+                    switch result {
+                    case let .success(value):
+                        decodedResponse = value
+                    case let .failure(error):
+                        decodingError = error
+                    }
+                case let .complete(completion):
+                    secondCompleteOnMain = Thread.isMainThread
+                    secondResponse = completion.response
+                    secondExpectation.fulfill()
+                }
+            }
+
+        waitForExpectations(timeout: timeout)
+
+        // Then
+        XCTAssertTrue(firstStreamOnMain)
+        XCTAssertTrue(firstCompleteOnMain)
+        XCTAssertEqual(firstResponse?.statusCode, 200)
+        XCTAssertTrue(!firstAccumulatedData.isEmpty)
+        XCTAssertTrue(secondStreamOnMain)
+        XCTAssertTrue(secondCompleteOnMain)
+        XCTAssertEqual(secondResponse?.statusCode, 200)
+        XCTAssertNotNil(decodedResponse)
+        XCTAssertNil(decodingError)
+    }
+
+    func testThatDataStreamRequestProducesWorkingInputStream() {
+        // Given
+        let expect = expectation(description: "stream complete")
+
+        // When
+        let stream = AF.streamRequest(URLRequest.makeHTTPBinRequest(path: "xml", headers: [.contentType("application/xml")]))
+            .responseStream { stream in
+                switch stream.event {
+                case .complete:
+                    expect.fulfill()
+                default: break
+                }
+            }.asInputStream()
```

Comment:
Will we crash if you don't put `responseStream` before `asInputStream`? All your docs do it in this order, but I'm not sure it's called out explicitly.

Reply:
It wouldn't crash, but it also wouldn't start the request. I've updated the logic in https://github.com/Alamofire/Alamofire/pull/3074/commits/52e4f454339fac1e5ad7009e631820f7c9aad107 to address that issue, please take a look.

Reply:
Looks good. The optional input stream is a bit clunky, but seems like a better tradeoff. I think the `defer { resume() }` makes sense as well.